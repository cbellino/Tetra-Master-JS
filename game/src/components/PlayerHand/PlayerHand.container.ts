import { connect } from "react-redux";

import { Id } from "../../models";
import {
  focusHandTile,
  getPlayer,
  isCurrentPlayerId,
  selectHandTile,
} from "../../store";
import { EnhancedPlayerHand } from "./PlayerHand";

type OwnProps = {
  playerId: Id;
};

const mapStateToProp = (state, ownProps: OwnProps) => {
  return {
    player: getPlayer(ownProps.playerId)(state),
    isPlaying: isCurrentPlayerId(ownProps.playerId)(state),
  };
};

const mapDispatchToProps = (dispatch, ownProps: OwnProps) => ({
  onTileMouseEnter: ({ tileId }) => {
    dispatch(focusHandTile(ownProps.playerId, tileId));
  },
  onTileMouseLeave: ({ tileId }) => {
    dispatch(focusHandTile(ownProps.playerId, null));
  },
  onTileClick: ({ tileId }) => {
    dispatch(selectHandTile(ownProps.playerId, tileId));
  },
});

const enhance = connect(
  mapStateToProp,
  mapDispatchToProps,
);

export default enhance(EnhancedPlayerHand);
