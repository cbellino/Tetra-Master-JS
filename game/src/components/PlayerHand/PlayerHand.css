.player-hand {
  position: absolute;
  top: 0;
  bottom: 50px;
  width: 100px;
  display: flex;
  flex-direction: column;
  padding: 0 32px;
}

.player-hand .player-name {
  min-height: 64px;
}

.player-hand .tile {
  margin-right: 8px;
}

.player-hand.player-1 {
  left: 0;
}

.player-hand.player-2 {
  right: 0;
}

.player-hand.is-playing .player-name:after {
  display: block;
  content: '(active)';
}
