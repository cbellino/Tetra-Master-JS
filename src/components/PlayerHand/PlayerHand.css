.player-hand {
  position: absolute;
  top: 0;
  width: 100px;
  display: flex;
  flex-direction: column;
  padding: 16px;
}

.player-hand .tile {
  margin-right: 8px;
}

.player-hand.player-1 {
  left: 0;
}

.player-hand.player-2 {
  right: 0;
}
