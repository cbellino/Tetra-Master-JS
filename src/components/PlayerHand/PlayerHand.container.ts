import { connect } from "react-redux";

import { Id } from "../../models";
import {
  focusHandTile,
  getPlayer,
  isCurrentPlayerId,
  selectHandTile,
} from "../../store";
import { EnhancedPlayerHand, EnhancedPlayerHandProps } from "./PlayerHand";

type OwnProps = {
  playerId: Id;
};
type MergeProps = EnhancedPlayerHandProps & OwnProps;

const mergeProps = (
  stateProps,
  dispatchProps,
  ownProps: OwnProps,
): MergeProps => {
  const { state } = stateProps;
  const { dispatch } = dispatchProps;
  const { playerId } = ownProps;

  const player = getPlayer(playerId)(state);
  const isPlaying = isCurrentPlayerId(player.id)(state);

  return {
    ...ownProps,
    player,
    isPlaying,
    onTileMouseEnter: ({ tileId }) => {
      dispatch(focusHandTile(playerId, tileId));
    },
    onTileMouseLeave: ({ tileId }) => {
      dispatch(focusHandTile(playerId, null));
    },
    onTileClick: ({ tileId }) => {
      dispatch(selectHandTile(playerId, tileId));
    },
  };
};

const enhance = connect(
  state => ({ state }),
  dispatch => ({ dispatch }),
  mergeProps,
);

export default enhance(EnhancedPlayerHand);
