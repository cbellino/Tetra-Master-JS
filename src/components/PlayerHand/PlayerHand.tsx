import classnames from "classnames";
import * as React from "react";
import { compose, defaultProps } from "recompose";

import { Id, Player } from "../../models";
import TileCard from "../TileCard";
import "./PlayerHand.css";

export type TileEventFn = (params: { tileId: Id }) => void;
export type EnhancedPlayerHandProps = {
  player: Player;
  isPlaying?: boolean;
  onTileMouseEnter?: TileEventFn;
  onTileMouseLeave?: TileEventFn;
  onTileClick?: TileEventFn;
};
type PlayerHandProps = {
  player: Player;
  isPlaying: boolean;
  onTileMouseEnter: TileEventFn;
  onTileMouseLeave: TileEventFn;
  onTileClick: TileEventFn;
};

const handleTileMouseEvent = (tileId: Id, onEvent: TileEventFn) => () => {
  onEvent({ tileId });
};

const PlayerHand: React.SFC<PlayerHandProps> = ({
  player,
  isPlaying,
  onTileMouseEnter,
  onTileMouseLeave,
  onTileClick,
}) => {
  const className = classnames("player-hand", `player-${player.id}`, {
    "is-playing": isPlaying,
  });

  return (
    <div className={className}>
      <h2 className="player-name">{player.name}</h2>
      {player.hand.map((tileId, index) => (
        <TileCard
          key={index}
          tileId={tileId}
          playerId={player.id}
          focused={tileId === player.focusedTileId}
          selected={tileId === player.selectedTileId}
          onMouseEnter={handleTileMouseEvent(tileId, onTileMouseEnter)}
          onMouseLeave={handleTileMouseEvent(tileId, onTileMouseLeave)}
          onClick={handleTileMouseEvent(tileId, onTileClick)}
        />
      ))}
    </div>
  );
};

const withDefaultProps = defaultProps({
  isPlaying: false,
  onTileMouseEnter: () => {},
  onTileMouseLeave: () => {},
  onTileClick: () => {},
});

const enhance = compose<PlayerHandProps, EnhancedPlayerHandProps>(
  withDefaultProps,
);

export const EnhancedPlayerHand = enhance(PlayerHand);
