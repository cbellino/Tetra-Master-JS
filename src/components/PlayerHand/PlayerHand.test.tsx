import { shallow } from "enzyme";
import * as React from "react";

import TileCard from "../TileCard";
import { EnhancedPlayerHand as PlayerHand } from "./PlayerHand";

const mockPlayer = { id: "1", name: "Player 1", hand: [] };
const mockPlayerWithHand = { id: "1", name: "Player 1", hand: ["1", "2"] };

describe("PlayerHand", () => {
  it("should display the player name ('Player 1')", () => {
    const wrapper = shallow(<PlayerHand player={mockPlayer} />).dive();

    expect(wrapper.find(".player-name").contains("Player 1")).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should have a className containing the player id", () => {
    const wrapper = shallow(<PlayerHand player={mockPlayer} />).dive();

    expect(wrapper.hasClass("player-1")).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  describe("when the player is playing", () => {
    it("should have a className of 'is-playing'", () => {
      const wrapper = shallow(
        <PlayerHand player={mockPlayer} isPlaying />,
      ).dive();

      expect(wrapper.hasClass("is-playing")).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("when the player has no tile in hand", () => {
    it("should display 0 tile", () => {
      const wrapper = shallow(<PlayerHand player={mockPlayer} />).dive();

      expect(wrapper.find(TileCard)).toHaveLength(0);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("when the player has 2 tiles in hand", () => {
    it("should display 2 tiles", () => {
      const wrapper = shallow(
        <PlayerHand player={mockPlayerWithHand} />,
      ).dive();

      expect(wrapper.find(TileCard)).toHaveLength(2);
      expect(wrapper).toMatchSnapshot();
    });

    it("should react to mouse events", () => {
      const onTileMouseEnter = jest.fn();
      const onTileMouseLeave = jest.fn();
      const onTileClick = jest.fn();
      const wrapper = shallow(
        <PlayerHand
          player={mockPlayerWithHand}
          onTileMouseEnter={onTileMouseEnter}
          onTileMouseLeave={onTileMouseLeave}
          onTileClick={onTileClick}
        />,
      ).dive();

      const firstTileCard = wrapper.find(TileCard).first();
      firstTileCard.simulate("click");
      expect(onTileClick).toHaveBeenCalledTimes(1);
      expect(onTileClick).toHaveBeenCalledWith({ tileId: "1" });

      firstTileCard.simulate("mouseEnter");
      expect(onTileMouseEnter).toHaveBeenCalledTimes(1);
      expect(onTileMouseEnter).toHaveBeenCalledWith({ tileId: "1" });

      firstTileCard.simulate("mouseLeave");
      expect(onTileMouseLeave).toHaveBeenCalledTimes(1);
      expect(onTileMouseLeave).toHaveBeenCalledWith({ tileId: "1" });

      expect(wrapper).toMatchSnapshot();
    });
  });
});
