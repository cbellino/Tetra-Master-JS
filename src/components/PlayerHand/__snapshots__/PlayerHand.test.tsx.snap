// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerHand should display the player name ('Player 1') 1`] = `
<div
  className="player-hand player-1"
>
  <h2
    className="player-name"
  >
    Player 1
  </h2>
</div>
`;

exports[`PlayerHand should have a className containing the player id 1`] = `
<div
  className="player-hand player-1"
>
  <h2
    className="player-name"
  >
    Player 1
  </h2>
</div>
`;

exports[`PlayerHand when the player has 2 tiles in hand should display 2 tiles 1`] = `
<div
  className="player-hand player-1"
>
  <h2
    className="player-name"
  >
    Player 1
  </h2>
  <Connect(TileCard)
    focused={false}
    key="0"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    playerId="1"
    selected={false}
    tileId="1"
  />
  <Connect(TileCard)
    focused={false}
    key="1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    playerId="1"
    selected={false}
    tileId="2"
  />
</div>
`;

exports[`PlayerHand when the player has 2 tiles in hand should react to mouse events 1`] = `
<div
  className="player-hand player-1"
>
  <h2
    className="player-name"
  >
    Player 1
  </h2>
  <Connect(TileCard)
    focused={false}
    key="0"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    playerId="1"
    selected={false}
    tileId="1"
  />
  <Connect(TileCard)
    focused={false}
    key="1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    playerId="1"
    selected={false}
    tileId="2"
  />
</div>
`;

exports[`PlayerHand when the player has no tile in hand should display 0 tile 1`] = `
<div
  className="player-hand player-1"
>
  <h2
    className="player-name"
  >
    Player 1
  </h2>
</div>
`;

exports[`PlayerHand when the player is playing should have a className of 'is-playing' 1`] = `
<div
  className="player-hand player-1 is-playing"
>
  <h2
    className="player-name"
  >
    Player 1
  </h2>
</div>
`;
